db_sql.go

```diff
@@ -123,22 +126,71 @@ func (x *sqlUserStoreDB) ResetPasswordFinish(userId UserId, token string, passwo
 			}
 			return escan
 		}
+				
 		if everify := verifyPasswordResetToken(token, truthToken.String); everify != nil {
 			tx.Rollback()
 			return everify
 		}
+
+		if used := x.verifyPasswordArchived(userId, password); used {
+			tx.Rollback()
+			return ErrInvalidPastPassword
+		}
+
+		if earchive := x.archivePassword(tx, userId); earchive != nil {
+			tx.Rollback()
+			return earchive
+		}
+
 		if eupdate := x.setPasswordInternal(tx, userId, password); eupdate != nil {
 			tx.Rollback()
 			return eupdate
-		} else {
-			return tx.Commit()
 		}
-
+		return tx.Commit()
 	} else {
 		return etx
 	}
 }
 
+func (x *sqlUserStoreDB) archivePassword(tx *sql.Tx, userId UserId) error {
+	if _, err := tx.Exec(`INSERT INTO authpwdarchive (userid, password) SELECT userid, password FROM authuserpwd WHERE userId = $1`, userId); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (x *sqlUserStoreDB) verifyPasswordArchived(userId UserId, password string) bool {
+	archivedPasswords, err := x.getArchivedPasswordsForUser(userId)
+	if err != nil {
+		return false
+	}
+	for _, hash := range archivedPasswords {
+		if verifyAuthausHash(password, hash) {
```

Comment:
A better way may be to calculate the hash first and then checking if it is in the archive table, instead of getting the last 15 passwords and checking each of them server side.

Reply:
@RoanBrand One actually needs to do it like this, because you can't compute the hash without the salt, and the salt is stored inside the blob that we refer to as the 'password hash'